### DISCUSSION TITLE
Weird MPI errors with distrubuted mesh

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#18603] [Q&A Meshing] [answered]
Hi. We are running a module on top of moose and use distributed meshes in order to avoid choking memory on the nodes. This has worked very well. Then the cluster was down for maintenance and after that we get weird errors - the code or build have not been touched. It may be that the errors and the maintenance are coincidental - I don't know. Anyhow, we do the usual - first run to distribute the mesh for N tasks, 
srun /lcrc/project/Meso/projects/hedgehog_kks/hedgehog_kks-opt -i KKS_CALPHAD_LHS_3D.i --split-mesh 144 --split-file foo.cpr

and then run the job
srun -n 144 /lcrc/project/Meso/projects/hedgehog_kks/hedgehog_kks-opt -i KKS_CALPHAD_LHS_3D.i --use-split --split-file foo.cpr
We use checkpointing and restarts. The errors are as below. Clearly weird stuff going on with MPI. The technical support people (who are quite good) are sure this has to do with the specific app and is something they cannot address directly. Any insight would be greatly appreciated.

Here is an example of the error:
[MPID_nem_tmi_pending_ssend_dequeue]: ERROR: can not find matching ssend. context=0, rank=65, tag=0
[MPID_nem_tmi_pending_ssend_dequeue]: ERROR: can not find matching ssend. context=0, rank=70, tag=0
[MPID_nem_tmi_pending_ssend_dequeue]: pending ssend list: (context=0, rank=0, tag=400008f5, sreq=0x4266be0) (context=0, rank=1, tag=400008f5, sreq=0x2b4436aafe80) (context=0, rank=4, tag=400008f5, sreq=0x2b4436aae800) (context=0, rank=5, tag=400008f5, sreq=0x2b4436ab5400) (context=0, rank=13, tag=400008f5, sreq=0x423b6e0) (context=0, rank=22, tag=400008f5, sreq=0x425a160) (context=0, rank=36, tag=400008f5, sreq=0x42338e0) (context=0, rank=48, tag=400008f5, sreq=0x2b4436abbb80) (context=0, rank=56, tag=400008f5, sreq=0x422bae0) (context=0, rank=63, tag=400008f5, sreq=0x42686e0) (context=0, rank=64, tag=400008f5, sreq=0x42302e0) (context=0, rank=65, tag=400008f5, sreq=0x425fb60) (context=0, rank=113, tag=400008f5, sreq=0x4239760) (context=0, rank=114, tag=400008f5, sreq=0x43652c0) (context=0, rank=118, tag=400008f5, sreq=0x4228960) (context=0, rank=130, tag=400008f5, sreq=0x43796c0)
[MPID_nem_tmi_pending_ssend_dequeue]: pending ssend list: (context=0, rank=0, tag=400008f5, sreq=0x3649c40) (context=0, rank=11, tag=400008f5, sreq=0x3634f40) (context=0, rank=15, tag=400008f5, sreq=0x3608cc0) (context=0, rank=57, tag=400008f5, sreq=0x3644b40) (context=0, rank=59, tag=400008f5, sreq=0x3631dc0) (context=0, rank=66, tag=400008f5, sreq=0x2b20f0d0e580) (context=0, rank=68, tag=400008f5, sreq=0x2b20f0d2a300) (context=0, rank=70, tag=400008f5, sreq=0x360d940) (context=0, rank=77, tag=400008f5, sreq=0x2b20f0d22e00) (context=0, rank=80, tag=400008f5, sreq=0x3632fc0) (context=0, rank=88, tag=400008f5, sreq=0x2b20f0d1ea80) (context=0, rank=99, tag=400008f5, sreq=0x2b20f0d20a00) (context=0, rank=101, tag=400008f5, sreq=0x2b20f0d16800)
MPID_nem_tmi_handle_ssend_ack: &pad1=0x2b4436b14300, &ssend_ack_recv_buffer=0x2b4436b14308, &pad2=0x2b4436b14318
MPID_nem_tmi_handle_ssend_ack: &pad1=0x2b20f0d7c300, &ssend_ack_recv_buffer=0x2b20f0d7c308, &pad2=0x2b20f0d7c318
MPID_nem_tmi_handle_ssend_ack: pad1=0, pad2=0 (both should be 0)
MPID_nem_tmi_handle_ssend_ack: pad1=0, pad2=0 (both should be 0)
srun: error: bdw-0605: task 124: Segmentation fault (core dumped)
srun: error: bdw-0531: task 98: Segmentation fault (core dumped)

**Comment by GiudGiud** [comment]
Hello

Did you re-compile after the cluster got updated? Did they change anything about the MPI distribution during maintenance?
Could we please get more information about the MOOSE version and the MPI distribution used?

Best,
Guillaume

**Comment by heinono1** [comment]
Hi,

I don’t think the MPI dist got updated (other mpi codes run without problems afaik). They did not do any software updates – they just had a problem with the file server that they sorted out.

-Olle

Olle Heinonen
Sr. Materials Scientist
***@***.***


From: Guillaume Giudicelli ***@***.***>
Reply-To: idaholab/moose ***@***.***>
Date: Sunday, August 15, 2021 at 7:40 PM
To: idaholab/moose ***@***.***>
Cc: "Heinonen, Olle G." ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)


Hello

Did you re-compile after the cluster got updated? Did they change anything about the MPI distribution?

Best,
Guillaume

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1187773>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF3JMHL25OPKM2QWL4TT5BM7LANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Reply by GiudGiud** [reply]
Does it work with less processes than 144?

**Reply by heinono1** [reply]
I haven't checked on fewer than 144 processes.
# moose-environment package version (https://github.com/idaholab/package_builder)
# https://github.com/idaholab/package_builder/pull/217
repo-hash:7af64c8a0de933a801d096655b79fb957ecc10b1

Intel compilers (17.0.4) and mpi wrappers, intel mkl

**Comment by heinono1** [comment]
I don’t think a bad node is the problem – we have run several times and gotten the same error. Because the cluster has some 1000 nodes, every time I queue up a job requesting 4 nodes I am likely to get 4 different nodes.

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Guillaume Giudicelli ***@***.***>
Date: Sunday, August 15, 2021 at 8:18 PM
To: idaholab/moose ***@***.***>
Cc: Heinonen, Olle G. ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)

Could it be that there is a bad node? Eg, does it work with less processes than 144?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1187857>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF6UL4PVXOIISK4N4TDT5BRNVANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Comment by GiudGiud** [comment]
Ok if recompiling does not help and if trying smaller configurations is not possible, we should try to get a backtrace for the error. 
Could you please recompile in debug mode with `METHOD=dbg make -j #procs`
then run inside a debugger with 
`srun -n 144 gdb --ex run --ex bt --args /lcrc/project/Meso/projects/hedgehog_kks/hedgehog_kks-dbg -i KKS_CALPHAD_LHS_3D.i --use-split --split-file foo.cpr`

**Comment by GiudGiud** [comment]
Any news of this? Has the problem been identified?

**Comment by heinono1** [comment]
It appears that something got mucked up in the mpi handling of distributed meshes. I have to rebuild moose….

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Guillaume Giudicelli ***@***.***>
Date: Tuesday, August 24, 2021 at 12:26 PM
To: idaholab/moose ***@***.***>
Cc: Heinonen, Olle G. ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)

Any news of this? Has the problem been identified?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1228751>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF2YVEUI4A57DOB4QRLT6PI3PANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Comment by heinono1** [comment]
Hi,

After battling conflicting versions of anaconda and miniconda, I did get the conda environment cleaned up and proceeded with the minconda3 install of moose on our large cluster (Ubuntu).  The build went without problem but the tests crash. A sample below. All crash errors (and pretty much everything crashes) seem to be related to problems in the actual codes or input files used for testing. What is going on???


(moose) ***@***.*** test]$ ./run_tests -j 4

outputs/vtk.solution/diff_parallel_mesh ................................. [#6149, MESH_MODE!=DISTRIBUTED] SKIP

outputs/png.image_tests/wedge ............................................................ [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/square_domain .................................................... [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/adv_diff_reaction ................................................ [LIBPNG!=TRUE] SKIP

outputs/nemesis.nemesis_scalar_distributed ..................................... [MESH_MODE!=DISTRIBUTED] SKIP

outputs/nemesis.nemesis_elemental_distributed .................................. [MESH_MODE!=DISTRIBUTED] SKIP

outputs/format.tecplot_bin_test_override ............................................... [TECPLOT!=FALSE] SKIP

controls/error.tid_warehouse_error ........................................................ [METHOD!=DBG] SKIP

time_integrators/explicit_ssp_runge_kutta.all/first_order: Working Directory: /gpfs/fs1/home/heinonen/projects/moose/test/tests/time_integrators/explicit_ssp_runge_kutta

time_integrators/explicit_ssp_runge_kutta.all/first_order: Running command: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt -i explicit_ssp_runge_kutta.i --error --error-unused --error-override --no-gdb-backtrace

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order: Exit Code: 127

time_integrators/explicit_ssp_runge_kutta.all/first_order: ################################################################################

time_integrators/explicit_ssp_runge_kutta.all/first_order: Tester failed, reason: CRASH

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order ..................................... FAILED (CRASH)


Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Tuesday, August 24, 2021 at 12:28 PM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
It appears that something got mucked up in the mpi handling of distributed meshes. I have to rebuild moose….

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Guillaume Giudicelli ***@***.***>
Date: Tuesday, August 24, 2021 at 12:26 PM
To: idaholab/moose ***@***.***>
Cc: Heinonen, Olle G. ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)

Any news of this? Has the problem been identified?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1228751>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF2YVEUI4A57DOB4QRLT6PI3PANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Comment by heinono1** [comment]
The main culprit seems to be that MPII_F_TRUE is undefined. Where is that supposed to be defined an what is the symbol for?

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Wednesday, August 25, 2021 at 8:25 AM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
Hi,

After battling conflicting versions of anaconda and miniconda, I did get the conda environment cleaned up and proceeded with the minconda3 install of moose on our large cluster (Ubuntu).  The build went without problem but the tests crash. A sample below. All crash errors (and pretty much everything crashes) seem to be related to problems in the actual codes or input files used for testing. What is going on???


(moose) ***@***.*** test]$ ./run_tests -j 4

outputs/vtk.solution/diff_parallel_mesh ................................. [#6149, MESH_MODE!=DISTRIBUTED] SKIP

outputs/png.image_tests/wedge ............................................................ [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/square_domain .................................................... [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/adv_diff_reaction ................................................ [LIBPNG!=TRUE] SKIP

outputs/nemesis.nemesis_scalar_distributed ..................................... [MESH_MODE!=DISTRIBUTED] SKIP

outputs/nemesis.nemesis_elemental_distributed .................................. [MESH_MODE!=DISTRIBUTED] SKIP

outputs/format.tecplot_bin_test_override ............................................... [TECPLOT!=FALSE] SKIP

controls/error.tid_warehouse_error ........................................................ [METHOD!=DBG] SKIP

time_integrators/explicit_ssp_runge_kutta.all/first_order: Working Directory: /gpfs/fs1/home/heinonen/projects/moose/test/tests/time_integrators/explicit_ssp_runge_kutta

time_integrators/explicit_ssp_runge_kutta.all/first_order: Running command: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt -i explicit_ssp_runge_kutta.i --error --error-unused --error-override --no-gdb-backtrace

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order: Exit Code: 127

time_integrators/explicit_ssp_runge_kutta.all/first_order: ################################################################################

time_integrators/explicit_ssp_runge_kutta.all/first_order: Tester failed, reason: CRASH

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order ..................................... FAILED (CRASH)


Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Tuesday, August 24, 2021 at 12:28 PM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
It appears that something got mucked up in the mpi handling of distributed meshes. I have to rebuild moose….

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Guillaume Giudicelli ***@***.***>
Date: Tuesday, August 24, 2021 at 12:26 PM
To: idaholab/moose ***@***.***>
Cc: Heinonen, Olle G. ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)

Any news of this? Has the problem been identified?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1228751>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF2YVEUI4A57DOB4QRLT6PI3PANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Reply by GiudGiud** [reply]
This seems like the mpi detected by the test suite and the mpi used to compile dont match

**Comment by heinono1** [comment, marked as answer]
OK, I had to purge the Lmod StdEnv and the tests seem to pass (at least they are still running without crashing).

I guess one thing I am curious about is performance comparisons between the gfortran and gcc (and their mpi wrappers) and the Intel compilers and wrappers.

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Wednesday, August 25, 2021 at 8:26 AM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
The main culprit seems to be that MPII_F_TRUE is undefined. Where is that supposed to be defined an what is the symbol for?

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Wednesday, August 25, 2021 at 8:25 AM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
Hi,

After battling conflicting versions of anaconda and miniconda, I did get the conda environment cleaned up and proceeded with the minconda3 install of moose on our large cluster (Ubuntu).  The build went without problem but the tests crash. A sample below. All crash errors (and pretty much everything crashes) seem to be related to problems in the actual codes or input files used for testing. What is going on???


(moose) ***@***.*** test]$ ./run_tests -j 4

outputs/vtk.solution/diff_parallel_mesh ................................. [#6149, MESH_MODE!=DISTRIBUTED] SKIP

outputs/png.image_tests/wedge ............................................................ [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/square_domain .................................................... [LIBPNG!=TRUE] SKIP

outputs/png.image_tests/adv_diff_reaction ................................................ [LIBPNG!=TRUE] SKIP

outputs/nemesis.nemesis_scalar_distributed ..................................... [MESH_MODE!=DISTRIBUTED] SKIP

outputs/nemesis.nemesis_elemental_distributed .................................. [MESH_MODE!=DISTRIBUTED] SKIP

outputs/format.tecplot_bin_test_override ............................................... [TECPLOT!=FALSE] SKIP

controls/error.tid_warehouse_error ........................................................ [METHOD!=DBG] SKIP

time_integrators/explicit_ssp_runge_kutta.all/first_order: Working Directory: /gpfs/fs1/home/heinonen/projects/moose/test/tests/time_integrators/explicit_ssp_runge_kutta

time_integrators/explicit_ssp_runge_kutta.all/first_order: Running command: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt -i explicit_ssp_runge_kutta.i --error --error-unused --error-override --no-gdb-backtrace

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order: /gpfs/fs1/home/heinonen/projects/moose/test/moose_test-opt: symbol lookup error: /home/heinonen/miniconda3/envs/moose/lib/./libmpifort.so.12: undefined symbol: MPII_F_TRUE

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order: Exit Code: 127

time_integrators/explicit_ssp_runge_kutta.all/first_order: ################################################################################

time_integrators/explicit_ssp_runge_kutta.all/first_order: Tester failed, reason: CRASH

time_integrators/explicit_ssp_runge_kutta.all/first_order:

time_integrators/explicit_ssp_runge_kutta.all/first_order ..................................... FAILED (CRASH)


Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Heinonen, Olle G. ***@***.***>
Date: Tuesday, August 24, 2021 at 12:28 PM
To: idaholab/moose ***@***.***>, idaholab/moose ***@***.***>
Cc: Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)
It appears that something got mucked up in the mpi handling of distributed meshes. I have to rebuild moose….

-Olle

Olle Heinonen
Sr. Materials Scientist
Argonne National Laboratory
***@***.***


From: Guillaume Giudicelli ***@***.***>
Date: Tuesday, August 24, 2021 at 12:26 PM
To: idaholab/moose ***@***.***>
Cc: Heinonen, Olle G. ***@***.***>, Author ***@***.***>
Subject: Re: [idaholab/moose] Weird MPI errors with distrubuted mesh (#18603)

Any news of this? Has the problem been identified?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/idaholab/moose/discussions/18603#discussioncomment-1228751>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEKZEF2YVEUI4A57DOB4QRLT6PI3PANCNFSM5CD35ZFA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>.

**Reply by GiudGiud** [reply]
It should be fairly similar. In my experience, threading is harder to get good performance with intel compilers, while they are slightly faster in serial.
